File: MyProjects/JavaFX/src/main/java/org/example/TaskManager.java
package org.example;

import java.io.*;
import java.util.*;
import java.util.logging.*;

public class TaskManager {
    private HashMap<Integer, Task> tasks = new HashMap<>();
    private int nextId = 1;
    private final String FILE_NAME = "tasks.dat";
    private static final Logger logger = Logger.getLogger(TaskManager.class.getName());

    public TaskManager() {
        loadFromFile();
    }

    public void addTask(String name, String description, int priority) {
        Task task = new Task(nextId++, name, description, priority);
        tasks.put(task.getId(), task);
        saveToFile();
    }

    public void removeTask(int id) {
        tasks.remove(id);
        saveToFile();
    }

    public void completeTask(int id) {
        Task t = tasks.get(id);
        if (t != null) {
            t.setCompleted(true);
            saveToFile();
        }
    }

    public List<Task> getAllTasks() {
        List<Task> list = new ArrayList<>(tasks.values());
        list.sort(Comparator.comparing(Task::getPriority));
        return list;
    }

    private void saveToFile() {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(FILE_NAME))) {
            oos.writeObject(tasks);
        } catch (IOException e) {
            logger.log(Level.SEVERE, "Fayla yazarkən xəta baş verdi", e);
        }
    }

    @SuppressWarnings("unchecked")
    private void loadFromFile() {
        File file = new File(FILE_NAME);
        if (!file.exists()) return;

        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(file))) {
            Object obj = ois.readObject();
            if (obj instanceof HashMap) {
                tasks = (HashMap<Integer, Task>) obj;
                if (!tasks.isEmpty())
                    nextId = Collections.max(tasks.keySet()) + 1;
            }
        } catch (IOException | ClassNotFoundException e) {
            logger.log(Level.WARNING, "Fayl oxunarkən xəta: " + e.getMessage(), e);
        }
    }
}

File: MyProjects/TestKotlin/src/main/kotlin/TaskManager.kt
package org.example

import java.io.*
import java.util.logging.Logger

class TaskManager {
    private var tasks = HashMap<Int, Task>()
    private var nextId = 1
    private val fileName = "tasks.dat"
    private val logger = Logger.getLogger(TaskManager::class.java.name)

    init {
        loadFromFile()
    }

    fun addTask(name: String, description: String, priority: Int) {
        val task = Task(nextId++, name, description, priority)
        tasks[task.id] = task
        saveToFile()
    }

    fun removeTask(id: Int) {
        tasks.remove(id)
        saveToFile()
    }

    fun completeTask(id: Int) {
        tasks[id]?.completed = true
        saveToFile()
    }

    fun getAllTasks(): List<Task> =
        tasks.values.sortedBy { it.priority }

    private fun saveToFile() {
        try {
            ObjectOutputStream(FileOutputStream(fileName)).use { it.writeObject(tasks) }
        } catch (e: IOException) {
            logger.severe("Fayla yazarkən xəta: ${e.message}")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadFromFile() {
        val file = File(fileName)
        if (!file.exists()) return

        try {
            ObjectInputStream(FileInputStream(file)).use { ois ->
                val obj = ois.readObject()
                if (obj is HashMap<*, *>) {
                    tasks = obj as HashMap<Int, Task>
                    if (tasks.isNotEmpty()) nextId = tasks.keys.maxOrNull()!! + 1
                }
            }
        } catch (e: Exception) {
            logger.warning("Fayl oxunarkən xəta: ${e.message}")
        }
    }
}
